File: algo/EvaluateReversePolishNotation.java
import java.util.Stack;

public class EvaluateReversePolishNotation {
    // Approach 1: Using Stack
    // Time Complexity: O(N)
    // Space Complexity: O(N)
    public int evalRPN(String[] tokens) {
        Stack<Integer> s = new Stack<>();
        for (int i = 0; i < tokens.length; i++) {

            if (tokens[i].equals("*")) {
                int nb2 = s.pop();
                int nb1 = s.pop();
                s.add(nb1 * nb2);
            } else if (tokens[i].equals("/")) {
                int nb2 = s.pop();
                int nb1 = s.pop();
                s.add(nb1 / nb2);
            } else if (tokens[i].equals("+")) {
                int nb2 = s.pop();
                int nb1 = s.pop();
                s.add(nb1 + nb2);
            } else if (tokens[i].equals("-")) {
                int nb2 = s.pop();
                int nb1 = s.pop();
                s.add(nb1 - nb2);
            } else {
                s.add(Integer.parseInt(tokens[i]));
            }
        }
        return s.peek();
    }
}

File: algo/EvaluateReversePolishNotation.cpp
class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        stack<long long int> stack;
        for (int i = 0; i < tokens.size(); i++)
        {
            string ch = tokens[i];
            if (ch.size() > 1 || isdigit(ch[0]))
            {
                stack.push(stoi(ch));
            }
            else
            {

                long long int num2 = stack.top();
                stack.pop();
                long long int num1 = stack.top();
                stack.pop();

                if (ch[0] == '*')
                {
                    stack.push(num1 * num2);
                }
                else if (ch[0] == '+')
                {
                    stack.push(num1 + num2);
                }
                else if (ch[0] == '-')
                {
                    stack.push(num1 - num2);
                }
                else if (ch[0] == '/')
                {
                    stack.push(num1 / num2);
                }
            }
        }
        return stack.top();
    }
};